rse-advanced-git.Rmd

```diff
@@ -0,0 +1,945 @@
+
+Now that we've merged the `pythondev` into `master`, these changes
+exist in both branches. This could be confusing in the future if we
+stumble upon the `pythondev` branch again.
+
+We can delete our old branches so as to avoid this confusion later.
+We can do so by adding the `-d` flag to the `git branch` command.
+
+```shell
+git branch -d pythondev
+```
+
+```text
+Deleted branch pythondev (was x792csa1).
+```
+
+And because we don't want to keep the changes in the `bashdev` branch,
+we can delete the `bashdev` branch as well
+
+```shell
+$ git branch -d bashdev
+```
+
+```text
+error: The branch 'bashdev' is not fully merged.
+If you are sure you want to delete it, run 'git branch -D bashdev'.
+```
+
+Since we've never merged the changes from the `bashdev` branch,
+Git warns us about deleting them and tells us to use the `-D` flag instead.
+
+Since we really want to delete this branch we will go ahead and do so.
+
+```shell
+git branch -D bashdev
+```
+
+```text
+Deleted branch bashdev (was 2n779ds).
+```
+
+
+### A branching workflow {#rse-advanced-git-branching-workflow}
+
+Now that we're comfortable with creating and merging branches,
+we need to consider when and how we might use branches in our code development workflows.
+
+Experienced developers tend to use a "branch-per-feature" workflow,
+even when they are working on their own.
+Whenever they want to create a new feature,
+they create a new branch from `master` and do the work there.
+When it's done,
+they merge that branch back into `master.
+
+In the context of what we've learned about branches so far,
+our branch-per-feature workflow looks like this:
+
+1.  `git checkout master` to make sure you're working in the `master` branch.
+
+2.  `git pull origin master` to make sure that
+    your starting point is up to date
+    (i.e., that the changes you committed while working on the cluster yesterday are now in your local copy,
+    or that the fix your colleague did last week has been incorporated).
+
+3.  `git checkout -b name-of-feature` to create a new branch.
+    The branch's name should be as descriptive as the name of a function
+    or of a file containing source code:
+    `"stuff"` or `"fixes"` saves you a couple of seconds of typing when you create the branch,
+    and will cost you much more time than that later on
+    when you're juggling half a dozen branches
+    and trying to figure out what's in each.
+
+4.  Do the work you want to do.
+    If some other task occurs to you while you're doing it---for example,
+    if you're writing a new function and realize that the documentation for some other function should be updated---*don't*
+    do that work in this branch just because you happen to be there.
+    Instead,
+    commit your changes,
+    switch back to `master`,
+    create a new branch for the other work,
+    and carry on.
+
+5.  Periodically,
+    you may `git push origin name-of-feature` to save your work to GitHub.
+
+6.  When the work is done,
```

Comment:
maybe worth being more explicit about the commands needed here -- that you need to be in master, pull, *then* switch to feature and merge? it's opposite of the example from before.